Implement slime blocks pushing and proper handling of non-movable blocks by pistons

Red-Teapot opened the pull request:
Added PistonMoveBehavior enum and corresponding methods in MaterialValueManager to store information about blocks push/pull behavior (e.g. bedrock does not push and pull, flowers drop when pushed and can't be pulled). All blocks can be pushed and pulled by default.

Pistons (including sticky variant) handle the behavior.

I haven't filled materialValues.yml yet so only bedrock and cobweb work for now.

Red-Teapot committed:
Implement PistonMoveBehavior

This enum specifies behavior of blocks that are going to be moved (pushed or pulled) by a piston.

Values are: MOVE, BREAK (e.g. flowers), DONT_MOVE (e.g. bedrock)

Red-Teapot committed:
Make pistons properly react to PistonMoveBehavior of blocks

Pr0methean commented on src/main/java/net/glowstone/block/PistonMoveBehavior.java:
```diff
@@ -0,0 +1,12 @@
+ * <p>MOVE - move as usual.</p>
+ * <p>BREAK - break block and drop necessary items (e.g. flowers).</p>
+ * <p>DONT_MOVE - don't move the block (e.g. bedrock).</p>
+ */
```
Doesn't the Slime Block need a fourth mode?

Red-Teapot replied:
What mode are you talking about? Special mode to determine if the block is slime? I don't know then, there is only one block with such properties for now.

And the Wiki says this:

> The types of blocks that can be moved are the same as those that can be pulled by a sticky piston. Blocks that cannot be pulled by a sticky piston (i.e. all the blocks listed on the table on the pistons page) stay in place.' 

So, I think that pistonPullBehavior property is enough for that.

Pr0methean replied:
I mean to implement the behavior where a slime block moves other blocks when moved by a piston. 
https://gamepedia.cursecdn.com/minecraft_gamepedia/9/94/SlimeBlockFlyingMachineAnimation300lrgr.gif?version=b8b2274f33e41ad881e1878965f314de

Red-Teapot replied:
Isn't it enough to just check if the material is slime block? So another variant in the enum is not needed.

I understand about the slime logic itself and I'm gonna work on it.

smartboyathome replied:
I would encourage you to rethink hard coding the slime block's "sticky" behavior. Even though slime blocks are the only instance of this behavior today doesn't mean it'll always be that way. Shouldn't be too much more work to abstract this behavior into some form of flag.

garantiertnicht replied:
Isn't there glazed terracotta as a special case that can only be pushed but not pulled?

mastercoms replied:
Isn't that covered by separate pull and push behavior?

Red-Teapot replied:
@garantiertnicht I've got separate move behaviors for pushing and pulling so it's not a problem. We can set Terracota's pushing behavior to MOVE and pulling behavior to DONT_MOVE.

Red-Teapot committed:
Implement dirty variant of pistons pushing slime blocks

aramperes commented on src/main/java/net/glowstone/block/blocktype/BlockPiston.java:
```diff
@@ -1,26 +1,41 @@
 
 public class BlockPiston extends BlockDirectional {
+    @AllArgsConstructor
+    private static class BlockInfo {
```
I think you can use the `@Data` annotation, which will be equivalent to AllArgsConstructor+G/S.

Pr0methean commented on src/main/java/net/glowstone/block/blocktype/BlockPiston.java:
```diff
@@ -186,6 +216,92 @@ public void onRedstoneUpdate(GlowBlock me) {
     }
 
+    // FIXME Remove this
+    private static Logger logger = Logger.getLogger(BlockPiston.class.getName());
```
Unused.

Pr0methean commented on src/main/java/net/glowstone/block/blocktype/BlockPiston.java:
```diff
@@ -1,26 +1,41 @@
+    private static class BlockInfo {
+        @Getter
+        @Setter
+        private GlowBlock block;
```
Why not just use these two fields as parameters instead of needing this inner class? It'll be simpler as long as there are only two or three of them.

Pr0methean commented on src/main/java/net/glowstone/block/blocktype/BlockPiston.java:
```diff
@@ -158,6 +216,92 @@ public void onRedstoneUpdate(GlowBlock me) {
+            }
+        }
+
+        if (block.getType() == Material.SLIME_BLOCK) {
```
Extract this into a subroutine. Add a PistonMoveBehavior.STICKY_MOVE and make it the value for Material.SLIME_BLOCK instead of checking for SLIME_BLOCK directly; that way, we can add the behavior to other blocks in the future.

Red-Teapot committed:
Get rid of unused Logger and BlockInfo, implement separate move behavior for slime block

Red-Teapot committed:
DRY block movement code a little

Red-Teapot committed:
Remove wildcard imports

Red-Teapot committed:
Add default cases to switches

Red-Teapot committed:
Format code

Red-Teapot committed:
Merge branch 'dev' into nonPushableBlocks

Red-Teapot committed:
Merge dev into nonPushableBlocks

Red-Teapot committed:
Remove unused Logger import in BlockPiston class

Red-Teapot commented:
I think I've fixed all the stuff I needed to fix. Material push/pull behaviors are not filled yet though. Works only for cobweb and bedrock just as an example of non-pushable/drop-on-move blocks.

aramperes commented on src/main/java/net/glowstone/block/blocktype/BlockPiston.java:
```diff
@@ -156,23 +181,104 @@ public void onRedstoneUpdate(GlowBlock me) {
 
-        // remove piston head after retracting
-        setType(me.getRelative(pistonBlockFace), 0, 0);
+    private void breakBlock(GlowBlock block) {
```
This logic doesn't take into account stuff like the `doTileDrops` gamerule and breaking containers (dropping chest contents). This kind of logic is already implemented in the [DiggingMessage handler](https://github.com/GlowstoneMC/Glowstone/blob/669ad8d45de9bea6ebe53a54bd6906fe9b4c5c87/src/main/java/net/glowstone/net/handler/play/player/DiggingHandler.java#L172-L185). Could you somehow merge this code?

If not, please leave a TODO and open an issue to support this.

Red-Teapot committed:
Merge branch 'dev' into nonPushableBlocks

Red-Teapot committed:
Implement gamerule check when a block is broken by a moving piston

Red-Teapot committed:
Fill in piston move behaviors for blocks

Based on https://minecraft.gamepedia.com/Piston#Usage
Entities (like item frames and paintings) not included.

Red-Teapot committed:
Fix piston movement dropping liquid items and empty item stacks (potentially)

Red-Teapot committed:
Fix code formatting

Pr0methean reviewed (commented):
What are the correct piston behaviors for water and lava? Also, do we need special handling for blocks that can fall after being pushed (sand, gravel, concrete powder, dragon egg)?

mastercoms commented:
Merging this because it's a great start in our incomplete server implementation. Will iterate further based on reviews in this PR.

Red-Teapot commented:
@Pr0methean Movement behaviors for liquids are specified in material values. The only difference is that `getMinedDrops()` for liquids returns item stack containing liquid block (the method is not overridden). I'm not sure if I should override that method to return an empty stack so I made a [workaround](https://github.com/Red-Teapot/Glowstone/blob/9019fd6d91080b13d25d572b2efada22043810f2/src/main/java/net/glowstone/block/blocktype/BlockPiston.java#L298). It seems like liquids are destroyed on push and do not move on pull in vanilla game.

Regarding gravity-affected stuff, it's hard to say because it is hardly working already, e.g. blocks don't fall when a block below them is destroyed. Maybe I'll try to find out how to make blocks fall when I get some free time.
